src/beast/include/beast/impl/basic_streambuf.ipp

```diff
@@ -478,74 +478,77 @@ auto
 basic_streambuf<Allocator>::prepare(size_type n) ->
     mutable_buffers_type
 {
-    iterator pos = out_;
-    if(pos != list_.end())
+    list_type reuse;
+    if(out_ != list_.end())
     {
-        auto const avail = pos->size() - out_pos_;
+        if(out_ != list_.iterator_to(list_.back()))
+        {
+            out_end_ = out_->size();
+            reuse.splice(reuse.end(), list_,
+                std::next(out_), list_.end());
+            debug_check();
+        }
+        auto const avail = out_->size() - out_pos_;
         if(n > avail)
         {
+            out_end_ = out_->size();
             n -= avail;
-            out_end_ = pos->size();
-            while(++pos != list_.end())
-            {
-                if(n < pos->size())
-                {
-                    out_end_ = n;
-                    n = 0;
-                    ++pos;
-                    break;
-                }
-                n -= pos->size();
-            }
         }
         else
         {
-            ++pos;
             out_end_ = out_pos_ + n;
             n = 0;
         }
+        debug_check();
     }
-
-    if(n > 0)
+    while(n > 0 && ! reuse.empty())
     {
-        assert(pos == list_.end());
-        for(;;)
+        auto& e = reuse.back();
```

Comment:
We're reusing buffers in reverse order from the original. I'd use `resue.front()` instead.